File: Python3/401 - 600/559 Maximum Depth of N-ary Tree.py
# 88ms, 60.29%; 17.4MB, 5.30%
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children

class Solution:
    def maxDepth(self, root: Node) -> int:
        if not root: return 0
        stack, depth = [root], 0
        while stack:
            depth += 1
            tmp = []
            while stack:
                ptr = stack.pop()
                for p in ptr.children:
                    tmp.append(p)
            stack = tmp
        return depth

File: Cpp/401 - 600/559 Maximum Depth of N-ary Tree.cpp
// 148ms, 98.91%; 33.7MB, 5.52%
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
static int x = []() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	return 0;
}();

class Solution
{
private:
	int depth = 0;

public:
	int maxDepth(Node *root)
	{
		if(!root) return depth;
		stack<Node *> st;
		st.push(root);
		while(!st.empty()) {
			++depth;
			stack<Node *> tmp;
			while(!st.empty()) {
				auto ptr = st.top();
				st.pop();
				for(auto p : ptr->children) tmp.push(p);
			}
			st.swap(tmp);
		}
		return depth;
	}
};
